File: spring-data-requery/src/main/java/org/springframework/data/requery/repository/query/RequeryQueryMethod.java
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.repository.query;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.core.annotation.AnnotatedElementUtils;
import org.springframework.core.annotation.AnnotationUtils;
import org.springframework.data.projection.ProjectionFactory;
import org.springframework.data.repository.core.RepositoryMetadata;
import org.springframework.data.repository.query.Parameter;
import org.springframework.data.repository.query.Parameters;
import org.springframework.data.repository.query.QueryMethod;
import org.springframework.data.requery.annotation.Query;
import org.springframework.util.Assert;
import org.springframework.util.StringUtils;

import java.lang.annotation.Annotation;
import java.lang.reflect.Method;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

/**
 * Requery specific extension of {@link QueryMethod}.
 *
 * @author debop
 * @since 18. 6. 7
 */
@Getter
@Slf4j
public class RequeryQueryMethod extends QueryMethod {

    private static final Set<Class<?>> NATIVE_ARRAY_TYPES;

    static {
        Set<Class<?>> types = new HashSet<>();
        types.add(byte[].class);
        types.add(Byte[].class);
        types.add(char[].class);
        types.add(Character[].class);

        NATIVE_ARRAY_TYPES = Collections.unmodifiableSet(types);
    }

    private final Method method;
    private final RepositoryMetadata metadata;
    private final RequeryEntityMetadata<?> entityInformation;

    /**
     * Creates a new {@link QueryMethod} from the given parameters. Looks up the correct query to use for following
     * invocations of the queryMethod given.
     *
     * @param method   must not be {@literal null}.
     * @param metadata must not be {@literal null}.
     * @param factory  must not be {@literal null}.
     */
    public RequeryQueryMethod(Method method, RepositoryMetadata metadata, ProjectionFactory factory) {
        super(method, metadata, factory);

        Assert.notNull(method, "Method must not be null!");

        log.debug("Create RequeryQueryMethod. repository={}, methodName={}, queryMethod={}",
                  metadata.getRepositoryInterface(), method.getName(), method);

        this.method = method;
        this.metadata = metadata;
        this.entityInformation = DefaultRequeryEntityMetadata.of(getDomainClass());

        Assert.isTrue(!(isModifyingQuery() && getParameters().hasSpecialParameter()),
                      String.format("Modifying queryMethod must not contains %s!", Parameters.TYPES));

        assertParamterNamesInAnnotatedQuery();
    }

    private void assertParamterNamesInAnnotatedQuery() {

        String annotatedQuery = getAnnotatedQuery();

        for (Parameter parameter : getParameters()) {
            if (!parameter.isNamedParameter()) {
                continue;
            }

            String paramName = parameter.getName().orElse("");
            if (StringUtils.isEmpty(annotatedQuery)
                || StringUtils.hasText(paramName)
                   && !annotatedQuery.contains(":" + paramName)
                   && !annotatedQuery.contains("#" + paramName)) {
                throw new IllegalStateException(
                    String.format("Using named parameters for queryMethod [%s] but parameter '%s' not found in annotated query '%s'!",
                                  method, parameter.getName(), annotatedQuery));
            }
        }
    }

    @NotNull
    @SuppressWarnings("unchecked")
    @Override
    public RequeryEntityMetadata<?> getEntityInformation() {
        return this.entityInformation;
    }

    @Override
    public boolean isModifyingQuery() {
        return false;
    }

    public boolean isAnnotatedQuery() {
        return null != AnnotationUtils.findAnnotation(method, Query.class);
    }

    public boolean isDefaultMethod() {
        return method.isDefault();
    }

    public boolean isOverridedMethod() {
        return !method.getDeclaringClass().equals(metadata.getRepositoryInterface());
    }

    Class<?> getReturnType() {
        return method.getReturnType();
    }

    @Nullable
    String getAnnotatedQuery() {
        String query = getAnnotationValue("value", String.class);
        return StringUtils.hasText(query) ? query : null;
    }

    String getRequiredAnnotatedQuery() throws IllegalStateException {
        String query = getAnnotatedQuery();
        if (query != null) {
            return query;
        }

        throw new IllegalStateException("No annotated query found for query queryMethod " + getName());
    }

    @Nullable
    String getCountQuery() {
        String countQuery = getAnnotationValue("countQuery", String.class);
        return StringUtils.hasText(countQuery) ? countQuery : null;
    }


    @SuppressWarnings("SameParameterValue")
    private String getAnnotationValue(String attribute, Class<String> type) {
        return getMergedOrDefaultAnnotationValue(attribute, Query.class, type);
    }

    @SuppressWarnings({ "unchecked", "SameParameterValue" })
    private <T> T getMergedOrDefaultAnnotationValue(String attribute,
                                                    Class annotationType,
                                                    Class<T> targetType) {
        Annotation annotation = AnnotatedElementUtils.findMergedAnnotation(method, annotationType);
        if (annotation == null) {
            return targetType.cast(AnnotationUtils.getDefaultValue(annotationType, attribute));
        }

        return targetType.cast(AnnotationUtils.getValue(annotation, attribute));
    }

    @NotNull
    @Override
    protected RequeryParameters createParameters(Method method) {
        return new RequeryParameters(method);
    }

    @Override
    public RequeryParameters getParameters() {
        return (RequeryParameters) super.getParameters();
    }

    @Override
    public boolean isCollectionQuery() {
        return super.isCollectionQuery() && !NATIVE_ARRAY_TYPES.contains(method.getReturnType());
    }

}

File: spring-data-requery-kotlin/src/main/kotlin/org/springframework/data/requery/kotlin/repository/query/RequeryQueryMethod.kt
/*
 * Copyright 2018 Coupang Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.springframework.data.requery.kotlin.repository.query

import mu.KotlinLogging
import org.springframework.core.annotation.AnnotatedElementUtils
import org.springframework.core.annotation.AnnotationUtils
import org.springframework.data.projection.ProjectionFactory
import org.springframework.data.repository.core.RepositoryMetadata
import org.springframework.data.repository.query.Parameters
import org.springframework.data.repository.query.QueryMethod
import org.springframework.data.requery.Annotations
import org.springframework.data.requery.kotlin.annotation.Query
import java.lang.reflect.Method
import kotlin.reflect.KClass

/**
 * RequeryQueryMethod
 *
 * @author debop
 */
class RequeryQueryMethod(val method: Method,
                         val metadata: RepositoryMetadata,
                         val factory: ProjectionFactory) : QueryMethod(method, metadata, factory) {

    companion object {
        private val log = KotlinLogging.logger { }

        // TODO: 혹시 이 것이 안되면, Java code에서 import 해야 한다.
        private val NATIVE_ARRAY_TYPES: Set<Class<*>> by lazy {
            LinkedHashSet<Class<*>>().apply {
                add(ByteArray::class.java)
                add(CharArray::class.java)
            }
        }
    }

    val entityInformation: RequeryEntityMetadata<*> = DefaultRequeryEntityMetadata.of(domainClass.kotlin)

    init {
        log.debug {
            "Create RequeryQueryMethod. repository=${metadata.repositoryInterface}, queryMethod name=${method.name}, queryMethod=$method"
        }
        check(!isModifyingQuery || !parameters.hasSpecialParameter()) {
            "Modifying query queryMethod must not contains ${Parameters.TYPES}"
        }

        assertParameterNamesInAnnotatedQuery()
    }

    private fun assertParameterNamesInAnnotatedQuery() {

        val query = annotatedQuery

        log.trace { "annotated query=$query" }
        /*
                parameters
                    .filter { it.isNamedParameter }
                    .forEach {

                        val paramName = it.name.orElse("")
                        log.trace { "check named parameter. paramName=$paramName" }

                        // Named Parameter 인데 Query에 ':paramName', '#paramName' 구문이 없다면 예외를 발생시킨다.
                        // NOTE: kotlin 함수의 일반 인자는 named parameter로 인식해버린다.


                        /*
                                        // NOTE: 현재 버전에서는 Named Parameter 자체를 지원하지 않으므로, 이 검사는 필요없다
                                        val notFoundParam = paramName.isNotBlank() &&
                                                            query != null &&
                                                            !query.contains(":$paramName") &&
                                                            !query.contains("#$paramName")
                                        if(notFoundParam) {
                                            error("Using named parameters for query queryMethod [$method] but parameter '${it.name}' not found in annotated query '$query'!")
                                        }
                        */
                    }
        */
    }


    val isAnnotatedQuery: Boolean
        get() = Annotations.findAnnotation(method, Query::class.java) != null

    val isDefaultMethod: Boolean
        get() = method.isDefault

    val isOverridedMethod: Boolean
        get() = method.declaringClass != metadata.repositoryInterface

    internal val returnType: Class<*>
        get() = method.returnType

    val annotatedQuery: String?
        get() {
            val query = getAnnotationValue("value", String::class.java)
            return when {
                query.isNullOrBlank() -> null
                else -> query
            }
        }

    val entityKlass: KClass<*>
        get() = entityInformation.javaType.kotlin

    val entityClass: Class<*>
        get() = entityInformation.javaType


    private fun getAnnotationValue(attribute: String, type: Class<String>): String? {
        return getMergedOrDefaultAnnotationValue(attribute, Query::class.java, type)
    }

    private fun <T> getMergedOrDefaultAnnotationValue(attribute: String,
                                                      annotationType: Class<out Annotation>,
                                                      targetType: Class<T>): T? {
        val annotation = AnnotatedElementUtils.findMergedAnnotation(method, annotationType)

        return annotation?.let {
            targetType.cast(AnnotationUtils.getValue(annotation, attribute))
        } ?: targetType.cast(AnnotationUtils.getDefaultValue(annotationType, attribute))
    }

    override fun createParameters(method: Method): Parameters<*, *> = RequeryParameters(method)

    override fun getParameters(): RequeryParameters = super.getParameters() as RequeryParameters

    override fun isCollectionQuery(): Boolean =
        super.isCollectionQuery() && !NATIVE_ARRAY_TYPES.contains(method.returnType)

}